### PR TITLE
feat: Add minimum first deposit, refactor convert/preview functions (SC-446)

### PR DESCRIPTION
**Pull request by lucas-manuel**
## Share Inflation Attack Initial Approach vs. Final Approach
This PR was originally created to address the issue outlined in picture below **within the contract's `deposit` logic**. This was fully implemented in [commit d32fd2f](https://github.com/marsfoundation/spark-psm/pull/2/commits/d32fd2faf5d3d255075f4eef9d6639f1118f5eb3). It was subsequently removed in favor of simplifying core logic in the contract and addressing the issue in a different way - by operationally requiring an initial deposit to be made on deployment. This would fix multiple attack vectors as outlined in @hexonaut's [comment](https://github.com/marsfoundation/spark-psm/pull/2#pullrequestreview-2085880206).

Screenshot 2024-05-16 at 3 51 13 PM

## Other Changes within PR

- Adds test to prove that the attack vector is resolved by the initial deposit being made.
- Adds section to README outlining attack prevention measures to be taken.
- Also refactors to use `preview` functions to return rounding-accurate conversions.
- Updates to have convertToShares and convertToAssets use generic value and asset-specific values.

### CODE REVIEW

**Review by hexonaut — CHANGES_REQUESTED** [review]
Mostly looks good. Some requested changes.

I would also make the recommendation to burn at least 1 full unit instead of 1000 shares.

File: src/PSM.sol [hunk 1]
```diff
@@ -103,36 +126,43 @@ contract PSM {
             ? assetBalance
             : maxAssetsToWithdraw;
 
-        uint256 sharesToBurn = convertToShares(asset, assetsWithdrawn);
+        sharesToBurn = _convertToSharesRoundUp(_getAssetValue(asset, assetsWithdrawn));
 
+        // TODO: Refactor this section to not use convertToAssets because of redundant check
+        // TODO: Can this cause an underflow in shares? Refactor to use full shares balance?
         if (sharesToBurn > shares[msg.sender]) {
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply src/PSM.sol:133]: Updated

File: src/PSM.sol [hunk 2]
```diff
@@ -94,6 +91,32 @@ contract PSM {
...
+        }
+
+        IERC20(asset).safeTransfer(msg.sender, assetsWithdrawn);
+    }
+
+    /**********************************************************************************************/
+    /*** Deposit/withdraw preview functions                                                     ***/
+    /**********************************************************************************************/
+
+    function previewDeposit(address asset, uint256 assets) public view returns (uint256) {
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply src/PSM.sol:111]: Updated

File: src/PSM.sol [hunk 3]
```diff
@@ -159,35 +197,27 @@ contract PSM {
...
-    function previewSwapAssetOneToZero(uint256 amountIn) public view returns (uint256) {
-        return amountIn
-            * IRateProviderLike(rateProvider).getConversionRate()
-            * asset0Precision
-            / 1e27
-            / asset1Precision;
+    function _divRoundUp(uint256 numerator_, uint256 divisor_)
+        internal pure returns (uint256 result_)
+    {
+        result_ = (numerator_ + divisor_ - 1) / divisor_;
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply src/PSM.sol:214]: Updated

**Review by hexonaut — APPROVED** [review]
Good to merge now.

**Pull request merged**